onnxruntime/core/providers/dnnl/subgraph/dnnl_convgrad.h

```diff
@@ -191,6 +191,30 @@ class DnnlConvGrad : public DnnlKernel {
     weights_md_ = onnxruntime::make_unique<dnnl::memory::desc>(
         dnnl::memory::desc({w_dims}, DnnnType<T>(), ort_source_format_));
 
+    // Verify that the inputs to the ConvGrad operator match the passed in forward Conv operator
+    // The Convolution forward operator is passed in from a stack durring the creation of the subgraph.
+    // There is an assumption that the ConvGrad nodes will be visited in the reverse order from the
+    // corresponding Conv Nodes. Unfortantly this is not guaranteed.
+    // It is possible that the push/pop order on the stack may endup incorrect.
+    // This checks the following:
+    //   - the output tensor dimensions from Conv match the dy_dims input of ConvGrad
+    //   - the input tensor dimensions from Conv match the x_dims input of ConvGrad
+    //   - the input weight tensor dimensions to Conv and ConvGrad match
+    // If any of these don't match then the stack did not assign the correct forward operator to this
+    // ConvGrad operator and the model will not be able to be trained using DNNL.
+    // See: dnnl_func_kernel.cc search for `fwd_conv_stack`.
+    // TODO: find a non-stack method to map the forward conv operator to the backward conv.
+    dnnl::memory::desc conv_fwd_dst_desc_ = conv_fwd_->GetPrimitiveDesc()->dst_desc();
+    dnnl::memory::desc conv_fwd_src_desc_ = conv_fwd_->GetPrimitiveDesc()->src_desc();
+    dnnl::memory::desc conv_fwd_weights_desc_ = conv_fwd_->GetPrimitiveDesc()->weights_desc();
+
+    if (conv_fwd_dst_desc_.dims() != dy_dims ||
```

Comment:
Is it possible to compare based on the *identity* of the input/weight tensors, rather than just their sizes? I'm not sure if this is feasible, and if not then adding a TODO would suffice. 

(If we happened to have two parallel convolutions that took the same sized tensors, we might inadvertently run an incorrect graph... it should be "safe" in the sense that all the buffers are the right size, but the output would be bogus.)

Reply:
I don't think I have access to the input/weight tensors _identity_. I will take a look. If I do have access to the _identity_ I may be able to remove the stack.

Reply:
@mrry do you know if the name of the input weight is unique to each input weight?  If so I could use the name of the weight to identify the tensors.

Reply:
@mrry in response to you asking me to see if it was possible to compare based on the _identity_ of the input/weight tensors I did some investigating. 

Originally I was unable to use the weight because the weight name was not filled in for ConvGrad node. I found the location in the code that the weight name should have been assigned for the ConvGrad and with that change I was able to remove the stack that was being used and replace it with a map.

This basically matches how I originally wanted to implement the code and your simple question about the _identity_ of the input/weight tensors helped me find the correct location in the code to make the change.

Thanks!

Reply:
That's wonderful news! Thanks for persevering with my awkward questions :).

Reply:
Your questions and requests did not but improve the code. Thanks for taking the time to review this huge commit.